File: scripts/extract_images/extract_images.py
##
 # LocalPDF Studio - Offline PDF Toolkit
 # ======================================
 # 
 # @author      Md. Alinur Hossain <alinur1160@gmail.com>
 # @license     MPL-2.0 (Mozilla Public License 2.0)
 # @website     https://alinur1.github.io/LocalPDF_Studio_Website/
 # @repository  https://github.com/Alinur1/LocalPDF_Studio
 # 
 # Copyright (c) 2025 Md. Alinur Hossain. All rights reserved.
 # 
 # This Source Code Form is subject to the terms of the Mozilla Public
 # License, v. 2.0. If a copy of the MPL was not distributed with this
 # file, You can obtain one at https://mozilla.org/MPL/2.0/.
 # 
 # Architecture:
 # - Frontend: Electron + HTML/CSS/JS
 # - Backend: ASP.NET Core Web API, Python
 # - PDF Engine: PdfSharp + Mozilla PDF.js
##


import fitz  # PyMuPDF
import json
import sys
import os
from PIL import Image
import io
import base64

def extract_images_from_pdf(pdf_path, pages=None, page_ranges=None, mode="extract"):
    """
    Extract or analyze images from PDF pages
    
    Args:
        pdf_path: Path to PDF file
        pages: List of specific page numbers (1-based)
        page_ranges: List of page ranges like ["1-3", "5-7"]
        mode: "extract" or "remove"
    
    Returns:
        Dictionary with results
    """
    try:
        doc = fitz.open(pdf_path)
        total_pages = doc.page_count
        
        # Determine pages to process
        pages_to_process = set()
        
        # If no pages specified, process all pages
        if not pages and not page_ranges:
            pages_to_process = set(range(total_pages))
        else:
            # Add specific pages
            if pages:
                for page_num in pages:
                    if 1 <= page_num <= total_pages:
                        pages_to_process.add(page_num - 1)  # Convert to 0-based
            
            # Add page ranges
            if page_ranges:
                for range_str in page_ranges:
                    if '-' in range_str:
                        start_str, end_str = range_str.split('-', 1)
                        try:
                            start = int(start_str.strip())
                            end = int(end_str.strip())
                            for page_num in range(start, end + 1):
                                if 1 <= page_num <= total_pages:
                                    pages_to_process.add(page_num - 1)
                        except ValueError:
                            continue
                    else:
                        # Single page in range format
                        try:
                            page_num = int(range_str.strip())
                            if 1 <= page_num <= total_pages:
                                pages_to_process.add(page_num - 1)
                        except ValueError:
                            continue
        
        pages_to_process = sorted(pages_to_process)
        
        if mode == "extract":
            return extract_images(doc, pages_to_process)
        else:  # remove mode
            return remove_images(doc, pages_to_process, pdf_path)
            
    except Exception as e:
        return {
            "success": False,
            "error": f"Error processing PDF: {str(e)}",
            "extracted_count": 0,
            "processed_pages": 0
        }
    finally:
        if 'doc' in locals():
            doc.close()

def extract_images(doc, pages_to_process):
    """Extract images from specified pages"""
    all_images = []
    total_images = 0
    
    for page_index in pages_to_process:
        page = doc[page_index]
        image_list = page.get_images()
        
        for img_index, img in enumerate(image_list):
            try:
                xref = img[0]
                pix = fitz.Pixmap(doc, xref)
                
                # Convert to RGB if needed
                if pix.n - pix.alpha < 4:  # can be saved as PNG
                    img_data = pix.tobytes("png")
                    
                    image_info = {
                        "page": page_index + 1,
                        "index": img_index,
                        "width": pix.width,
                        "height": pix.height,
                        "format": "png",
                        "data": base64.b64encode(img_data).decode('ascii')
                    }
                    all_images.append(image_info)
                    total_images += 1
                
                pix = None  # Free pixmap memory
                
            except Exception as e:
                print(f"Warning: Failed to extract image {img_index} from page {page_index + 1}: {e}", file=sys.stderr)
                continue
    
    return {
        "success": True,
        "extracted_count": total_images,
        "processed_pages": len(pages_to_process),
        "images": all_images
    }

def remove_images(doc, pages_to_process, original_path):
    """Remove images from specified pages and return modified PDF"""
    try:
        # Create a new document
        new_doc = fitz.open()
        
        # Copy all pages from original to new document
        new_doc.insert_pdf(doc)
        
        images_removed_count = 0
        
        # Remove images from specified pages in the new document
        for page_index in pages_to_process:
            if page_index < len(new_doc):
                page = new_doc[page_index]
                image_list = page.get_images()
                
                for img in image_list:
                    xref = img[0]
                    try:
                        # Remove the image object from the PDF
                        new_doc._deleteObject(xref)
                        images_removed_count += 1
                    except Exception as e:
                        print(f"Warning: Could not remove image xref {xref} from page {page_index + 1}: {e}", file=sys.stderr)
                        continue
        
        # Save to a bytes buffer
        import io
        pdf_buffer = io.BytesIO()
        new_doc.save(pdf_buffer)
        pdf_data = pdf_buffer.getvalue()
        new_doc.close()
        
        if not pdf_data:
            return {
                "success": False,
                "error": "Failed to generate PDF data",
                "processed_pages": 0
            }
        
        return {
            "success": True,
            "processed_pages": len(pages_to_process),
            "pdf_data": base64.b64encode(pdf_data).decode('ascii'),
            "removed_images_count": images_removed_count
        }
        
    except Exception as e:
        # Ensure documents are closed in case of error
        if 'new_doc' in locals():
            new_doc.close()
        return {
            "success": False,
            "error": f"Error removing images: {str(e)}",
            "processed_pages": 0
        }

def main():
    if len(sys.argv) < 2:
        error_result = {"success": False, "error": "No arguments provided"}
        print(json.dumps(error_result))
        sys.exit(1)
    
    try:
        # Read JSON from file (first argument is file path)
        json_file_path = sys.argv[1]
        
        with open(json_file_path, 'r', encoding='utf-8') as f:
            request = json.load(f)
        
        pdf_path = request.get("file_path")
        pages = request.get("pages")
        page_ranges = request.get("page_ranges")
        mode = request.get("mode", "extract")
        
        if not pdf_path or not os.path.exists(pdf_path):
            error_result = {"success": False, "error": f"PDF file not found: {pdf_path}"}
            print(json.dumps(error_result))
            sys.exit(1)
        
        result = extract_images_from_pdf(pdf_path, pages, page_ranges, mode)
        print(json.dumps(result))
        
    except json.JSONDecodeError as e:
        error_result = {"success": False, "error": f"Invalid JSON input: {str(e)}"}
        print(json.dumps(error_result))
        sys.exit(1)
    except Exception as e:
        error_result = {"success": False, "error": f"Processing error: {str(e)}"}
        print(json.dumps(error_result))
        sys.exit(1)

if __name__ == "__main__":
    main()

File: src/renderer/tools/extractImages/extractImages.js
/**
 * LocalPDF Studio - Offline PDF Toolkit
 * ======================================
 * 
 * @author      Md. Alinur Hossain <alinur1160@gmail.com>
 * @license     AGPL 3.0 (GNU Affero General Public License version 3)
 * @website     https://alinur1.github.io/LocalPDF_Studio_Website/
 * @repository  https://github.com/Alinur1/LocalPDF_Studio
 * 
 * Copyright (c) 2025 Md. Alinur Hossain. All rights reserved.
 * 
 * Architecture:
 * - Frontend: Electron + HTML/CSS/JS
 * - Backend: ASP.NET Core Web API, Python
 * - PDF Engine: PdfSharp + Mozilla PDF.js
**/

//src/renderer/tools/extractImages/extractImages.js

import * as pdfjsLib from '../../../pdf/build/pdf.mjs';
import { API } from '../../api/api.js';
import customAlert from '../../utils/customAlert.js';
import loadingUI from '../../utils/loading.js';

pdfjsLib.GlobalWorkerOptions.workerSrc = '../../../pdf/build/pdf.worker.mjs';
window.pdfjsLib = pdfjsLib;

document.addEventListener('DOMContentLoaded', async () => {
    await API.init();

    const selectPdfBtn = document.getElementById('select-pdf-btn');
    const removePdfBtn = document.getElementById('remove-pdf-btn');
    const processBtn = document.getElementById('process-btn');
    const previewBtn = document.getElementById('preview-btn');
    const selectedFileInfo = document.getElementById('selected-file-info');
    const pdfNameEl = document.getElementById('pdf-name');
    const pdfSizeEl = document.getElementById('pdf-size');
    const previewContainer = document.getElementById('preview-container');
    const previewGrid = document.getElementById('preview-grid');
    const pageCountEl = document.getElementById('page-count');
    const selectionInfoEl = document.getElementById('selection-info');
    const clearSelectionBtn = document.getElementById('clear-selection-btn');
    const operationModeRadios = document.querySelectorAll('input[name="operation-mode"]');
    const extractInfo = document.getElementById('extract-info');
    const removeInfo = document.getElementById('remove-info');
    const extractOptions = document.getElementById('extract-options');
    const selectAllBtn = document.getElementById('select-all-pages');
    const selectEvenBtn = document.getElementById('select-even-pages');
    const selectOddBtn = document.getElementById('select-odd-pages');
    const invertSelectionBtn = document.getElementById('invert-selection');
    const manualPagesInput = document.getElementById('manual-pages');
    const pageRangesInput = document.getElementById('page-ranges');
    const preserveQualityCheckbox = document.getElementById('preserve-quality');
    const preserveFormatCheckbox = document.getElementById('preserve-format');
    const preserveMetadataCheckbox = document.getElementById('preserve-metadata');
    let selectedFile = null;
    let pdfDoc = null;
    let renderedPages = [];
    let selectedPages = new Set();
    let totalPages = 0;
    let currentMode = 'extract';

    selectPdfBtn.addEventListener('click', async () => {
        loadingUI.show("Selecting PDF files...");
        const files = await window.electronAPI.selectPdfs();
        if (files && files.length > 0) {
            const filePath = files[0];
            const fileName = filePath.split(/[\\/]/).pop();
            const fileSize = await getFileSize(filePath);
            handleFileSelected({ path: filePath, name: fileName, size: fileSize });
        }
        loadingUI.hide();
    });

    removePdfBtn.addEventListener('click', () => clearAll());

    const backBtn = document.querySelector('a[href="../../index.html"]');
    if (backBtn) {
        backBtn.addEventListener('click', (e) => {
            e.preventDefault();
            clearAll();
            window.location.href = '../../index.html';
        });
    }

    async function handleFileSelected(file) {
        clearAll();
        selectedFile = file;
        pdfNameEl.textContent = file.name;
        pdfSizeEl.textContent = `(${(file.size / 1024 / 1024).toFixed(2)} MB)`;
        selectPdfBtn.style.display = 'none';
        selectedFileInfo.style.display = 'flex';
        await loadPdfPreview(file.path);
        updateButtonStates();
    }

    async function loadPdfPreview(filePath) {
        try {
            loadingUI.show('Loading PDF preview...');
            previewContainer.style.display = 'block';
            previewGrid.innerHTML = '';
            const loadingTask = pdfjsLib.getDocument(`file://${filePath}`);
            pdfDoc = await loadingTask.promise;
            totalPages = pdfDoc.numPages;
            pageCountEl.textContent = `Total Pages: ${totalPages}`;
            previewGrid.innerHTML = '';
            for (let pageNum = 1; pageNum <= totalPages; pageNum++) {
                await renderPageThumbnail(pageNum);
            }
        } catch (error) {
            console.error('Error loading PDF:', error);
            previewGrid.innerHTML = '<p style="color: #e74c3c; text-align: center;">Failed to load PDF preview</p>';
        } finally {
            loadingUI.hide();
        }
    }

    async function renderPageThumbnail(pageNum) {
        const page = await pdfDoc.getPage(pageNum);
        const scale = 0.3;
        const viewport = page.getViewport({ scale });
        const canvas = document.createElement('canvas');
        const context = canvas.getContext('2d');
        canvas.height = viewport.height;
        canvas.width = viewport.width;
        await page.render({ canvasContext: context, viewport }).promise;
        const thumbWrapper = document.createElement('div');
        thumbWrapper.className = 'page-thumbnail';
        thumbWrapper.dataset.pageNum = pageNum;
        const pageLabel = document.createElement('div');
        pageLabel.className = 'page-label';
        pageLabel.textContent = `Page ${pageNum}`;
        thumbWrapper.appendChild(canvas);
        thumbWrapper.appendChild(pageLabel);
        thumbWrapper.addEventListener('click', () => {
            togglePageSelection(pageNum, thumbWrapper);
        });
        previewGrid.appendChild(thumbWrapper);
        renderedPages.push(canvas);
    }

    function togglePageSelection(pageNum, element) {
        if (selectedPages.has(pageNum)) {
            selectedPages.delete(pageNum);
            element.classList.remove('selected');
        } else {
            selectedPages.add(pageNum);
            element.classList.add('selected');
        }
        updateSelectionInfo();
        updateButtonStates();
    }

    function updateSelectionInfo() {
        const count = selectedPages.size;
        if (count === 0) {
            selectionInfoEl.textContent = 'No pages selected';
            clearSelectionBtn.style.display = 'none';
        } else {
            const sortedPages = Array.from(selectedPages).sort((a, b) => a - b);
            if (count <= 10) {
                selectionInfoEl.textContent = `Selected: ${sortedPages.join(', ')} (${count} page${count > 1 ? 's' : ''})`;
            } else {
                selectionInfoEl.textContent = `${count} pages selected`;
            }
            clearSelectionBtn.style.display = 'block';
        }
    }

    clearSelectionBtn.addEventListener('click', () => {
        selectedPages.clear();
        document.querySelectorAll('.page-thumbnail').forEach(thumb => {
            thumb.classList.remove('selected');
        });
        updateSelectionInfo();
        updateButtonStates();
    });

    selectAllBtn.addEventListener('click', () => {
        for (let i = 1; i <= totalPages; i++) {
            selectedPages.add(i);
            const thumb = document.querySelector(`.page-thumbnail[data-page-num="${i}"]`);
            if (thumb) thumb.classList.add('selected');
        }
        updateSelectionInfo();
        updateButtonStates();
    });

    selectEvenBtn.addEventListener('click', () => {
        for (let i = 2; i <= totalPages; i += 2) {
            selectedPages.add(i);
            const thumb = document.querySelector(`.page-thumbnail[data-page-num="${i}"]`);
            if (thumb) thumb.classList.add('selected');
        }
        updateSelectionInfo();
        updateButtonStates();
    });

    selectOddBtn.addEventListener('click', () => {
        for (let i = 1; i <= totalPages; i += 2) {
            selectedPages.add(i);
            const thumb = document.querySelector(`.page-thumbnail[data-page-num="${i}"]`);
            if (thumb) thumb.classList.add('selected');
        }
        updateSelectionInfo();
        updateButtonStates();
    });

    invertSelectionBtn.addEventListener('click', () => {
        const newSelection = new Set();
        for (let i = 1; i <= totalPages; i++) {
            if (!selectedPages.has(i)) {
                newSelection.add(i);
            }
        }
        selectedPages = newSelection;

        document.querySelectorAll('.page-thumbnail').forEach(thumb => {
            const pageNum = parseInt(thumb.dataset.pageNum);
            if (selectedPages.has(pageNum)) {
                thumb.classList.add('selected');
            } else {
                thumb.classList.remove('selected');
            }
        });
        updateSelectionInfo();
        updateButtonStates();
    });

    operationModeRadios.forEach(radio => {
        radio.addEventListener('change', (e) => {
            currentMode = e.target.value;
            updateModeUI();
            updateButtonStates();
        });
    });

    function updateModeUI() {
        if (currentMode === 'extract') {
            extractInfo.style.display = 'flex';
            removeInfo.style.display = 'none';
            extractOptions.style.display = 'block';
            processBtn.textContent = 'Extract Images';
        } else {
            extractInfo.style.display = 'none';
            removeInfo.style.display = 'flex';
            extractOptions.style.display = 'none';
            processBtn.textContent = 'Remove Images';
        }
    }

    [manualPagesInput, pageRangesInput].forEach(input => {
        input.addEventListener('input', updateButtonStates);
        input.addEventListener('change', updateButtonStates);
    });

    function updateButtonStates() {
        const hasFile = selectedFile !== null;
        const hasSelection = selectedPages.size > 0 ||
            manualPagesInput.value.trim() !== '' ||
            pageRangesInput.value.trim() !== '';

        processBtn.disabled = !hasFile || !hasSelection;
        previewBtn.disabled = !hasFile || !hasSelection;
    }

    previewBtn.addEventListener('click', () => {
        const pagesToProcess = collectPagesToProcess();
        if (pagesToProcess.size === 0) {
            customAlert.alert('LocalPDF Studio - NOTICE', 'No pages selected.', ['OK']);
            return;
        }
        const sortedPages = Array.from(pagesToProcess).sort((a, b) => a - b);
        const modeText = currentMode === 'extract' ? 'extract images from' : 'remove images from';
        customAlert.alert('LocalPDF Studio - NOTICE', `Preview:\n\nPages to ${modeText}: ${sortedPages.join(', ')}\nTotal pages: ${pagesToProcess.size}`, ['OK']);
    });

    processBtn.addEventListener('click', async () => {
        if (!selectedFile) {
            await customAlert.alert('LocalPDF Studio - NOTICE', 'Please select a file first.', ['OK']);
            return;
        }

        const pagesToProcess = collectPagesToProcess();

        if (pagesToProcess.size === 0) {
            await customAlert.alert('LocalPDF Studio - NOTICE', 'Please select at least one page.', ['OK']);
            return;
        }

        const options = buildProcessOptions(pagesToProcess);

        const requestBody = {
            filePath: selectedFile.path,
            options: options
        };

        try {
            const loadingMessage = currentMode === 'extract' ? 'Extracting images...' : 'Removing images...';
            loadingUI.show(loadingMessage);
            processBtn.disabled = true;
            const originalText = processBtn.textContent;
            processBtn.textContent = currentMode === 'extract' ? 'Extracting...' : 'Removing...';

            let endpoint;
            if (currentMode === 'extract') {
                endpoint = await API.pdf.extractImages;
            } else {
                endpoint = await API.pdf.removeImages;
            }

            const result = await API.request.post(endpoint, requestBody);
            if (result instanceof Blob) {
                const arrayBuffer = await result.arrayBuffer();
                if (currentMode === 'extract') {
                    const defaultName = `${selectedFile.name.replace('.pdf', '')}_extracted_images.zip`;
                    const savedPath = await window.electronAPI.saveZipFile(defaultName, arrayBuffer);

                    if (savedPath) {
                        await customAlert.alert('LocalPDF Studio - SUCCESS', `Success! Images extracted successfully!\nSaved to: ${savedPath}`, ['OK']);
                    } else {
                        await customAlert.alert('LocalPDF Studio - WARNING', 'Operation cancelled or failed to save the file.', ['OK']);
                    }
                } else {
                    const defaultName = `${selectedFile.name.replace('.pdf', '')}_images_removed.pdf`;
                    const savedPath = await window.electronAPI.savePdfFile(defaultName, arrayBuffer);

                    if (savedPath) {
                        await customAlert.alert('LocalPDF Studio - SUCCESS', `Success! Images removed successfully!\nSaved to: ${savedPath}`, ['OK']);
                    } else {
                        await customAlert.alert('LocalPDF Studio - WARNING', 'Operation cancelled or failed to save the file.', ['OK']);
                    }
                }
            } else {
                console.error("API returned JSON:", result);
                await customAlert.alert('LocalPDF Studio - ERROR', `Error: ${JSON.stringify(result)}`, ['OK']);
            }
        } catch (error) {            
            console.error('Error processing images:', error);
            await customAlert.alert('LocalPDF Studio - ERROR', `An error occurred.`, ['OK']);
        } finally {
            loadingUI.hide();
            processBtn.disabled = false;
            processBtn.textContent = currentMode === 'extract' ? 'Extract Images' : 'Remove Images';
        }
    });

    function collectPagesToProcess() {
        const pagesToProcess = new Set(selectedPages);

        const manualPages = manualPagesInput.value.trim();
        if (manualPages) {
            const pages = manualPages.split(',').map(p => parseInt(p.trim())).filter(p => !isNaN(p) && p >= 1 && p <= totalPages);
            pages.forEach(p => pagesToProcess.add(p));
        }

        const ranges = pageRangesInput.value.trim();
        if (ranges) {
            const rangeList = ranges.split(',').map(r => r.trim());
            rangeList.forEach(range => {
                const parts = range.split('-');
                if (parts.length === 2) {
                    const start = parseInt(parts[0].trim());
                    const end = parseInt(parts[1].trim());
                    if (!isNaN(start) && !isNaN(end) && start >= 1 && end <= totalPages && start <= end) {
                        for (let i = start; i <= end; i++) {
                            pagesToProcess.add(i);
                        }
                    }
                }
            });
        }

        return pagesToProcess;
    }

    function buildProcessOptions(pagesToProcess) {
        const options = {
            mode: currentMode
        };

        const pagesArray = Array.from(pagesToProcess).sort((a, b) => a - b);

        const ranges = [];
        const individualPages = [];

        let rangeStart = null;
        let rangeEnd = null;

        for (let i = 0; i < pagesArray.length; i++) {
            const currentPage = pagesArray[i];

            if (rangeStart === null) {
                rangeStart = currentPage;
                rangeEnd = currentPage;
            } else if (currentPage === rangeEnd + 1) {
                rangeEnd = currentPage;
            } else {
                if (rangeEnd - rangeStart >= 2) {
                    ranges.push(`${rangeStart}-${rangeEnd}`);
                } else {
                    for (let j = rangeStart; j <= rangeEnd; j++) {
                        individualPages.push(j);
                    }
                }
                rangeStart = currentPage;
                rangeEnd = currentPage;
            }
        }

        if (rangeStart !== null) {
            if (rangeEnd - rangeStart >= 2) {
                ranges.push(`${rangeStart}-${rangeEnd}`);
            } else {
                for (let j = rangeStart; j <= rangeEnd; j++) {
                    individualPages.push(j);
                }
            }
        }

        if (individualPages.length > 0) {
            options.pages = individualPages;
        }

        if (ranges.length > 0) {
            options.pageRanges = ranges;
        }

        if (currentMode === 'extract') {
            options.preserveQuality = preserveQualityCheckbox.checked;
            options.preserveFormat = preserveFormatCheckbox.checked;
            options.preserveMetadata = preserveMetadataCheckbox.checked;
        }

        return options;
    }

    function clearAll() {
        if (pdfDoc) {
            pdfDoc.destroy();
            pdfDoc = null;
        }
        renderedPages.forEach(c => {
            const ctx = c.getContext('2d');
            ctx.clearRect(0, 0, c.width, c.height);
        });
        renderedPages = [];
        previewGrid.innerHTML = '';
        previewContainer.style.display = 'none';
        selectedFile = null;
        selectedPages.clear();
        totalPages = 0;
        selectedFileInfo.style.display = 'none';
        selectPdfBtn.style.display = 'block';

        manualPagesInput.value = '';
        pageRangesInput.value = '';

        updateButtonStates();
        updateSelectionInfo();
    }

    async function getFileSize(filePath) {
        try {
            if (window.electronAPI?.getFileInfo) {
                const info = await window.electronAPI.getFileInfo(filePath);
                return info.size || 0;
            }
            return 0;
        } catch {
            return 0;
        }
    }
    updateModeUI();
});
